### ISSUE TITLE
pinentry-mac causes `nix-store --optimise` to fail

### ISSUE THREAD

**Issue by clacke** [NixOS/nixpkgs#25819]
## Issue description

When I run `nix-store --optimise`, it fails when it encounters pinentry-mac, and I believe it returns immediately and doesn't optimize the rest of the store.

### Steps to reproduce

```bash
$ nix-store --optimise
0.00 MiB freed by hard-linking 0 files
$ nix-build -A pkgs.pinentry_mac '<nixpkgs>'
[ . . . ]
/nix/store/diw4mj51kw8njrv36ah7b3bjmkmvamph-pinentry-mac-0.9.4
$ nix-store --optimise
error: cannot link ‘/nix/store/.links/1666wf33g20gqagij48gmh2wk7vxq07kp0a0avqf13hyjfyii7wx’ to ‘/nix/store/diw4mj51kw8njrv36ah7b3bjmkmvamph-pinentry-mac-0.9.4/Applications/pinentry-mac.app/Contents/PkgInfo’: Operation not permitted
```
## Technical details

* System: macOS 10.12.4 Sierra
* Nix version: nix-env (Nix) 1.11.6
* Nixpkgs version: 17.09pre107260.e33848568d
* Sandboxing enabled: no

**Comment by LnL7** [comment]
Can you reproduce this if you delete the store path and rebuild it?

**Comment by matthewbauer** [comment]
This is probably an SIP issue. Something to try is:

```
chflags nouchg /nix/store/diw4mj51kw8njrv36ah7b3bjmkmvamph-pinentry-mac-0.9.4/Applications/pinentry-mac.app/Contents/PkgInfo
```

**Comment by clacke** [comment]
@LnL7 Yes, I removed it, optimizing was fine, built it again, problem resurfaced. Tried it a few weeks apart with different versions.

**Comment by clacke** [comment]
No flags anywhere, still fails:

```
$ /bin/ls -lOd /nix/store/diw4mj51kw8njrv36ah7b3bjmkmvamph-pinentry-mac-0.9.4{,/Applications{,/pinentry-mac.app{,/Contents{,/PkgInfo}}}}
dr-xr-xr-x  3 root  nixbld  - 102 Jan  1  1970 /nix/store/diw4mj51kw8njrv36ah7b3bjmkmvamph-pinentry-mac-0.9.4
dr-xr-xr-x  3 root  nixbld  - 102 Jan  1  1970 /nix/store/diw4mj51kw8njrv36ah7b3bjmkmvamph-pinentry-mac-0.9.4/Applications
dr-xr-xr-x  3 root  nixbld  - 102 Jan  1  1970 /nix/store/diw4mj51kw8njrv36ah7b3bjmkmvamph-pinentry-mac-0.9.4/Applications/pinentry-mac.app
dr-xr-xr-x  6 root  nixbld  - 204 Jan  1  1970 /nix/store/diw4mj51kw8njrv36ah7b3bjmkmvamph-pinentry-mac-0.9.4/Applications/pinentry-mac.app/Contents
-r--r--r--  1 root  nixbld  -   8 Jan  1  1970 /nix/store/diw4mj51kw8njrv36ah7b3bjmkmvamph-pinentry-mac-0.9.4/Applications/pinentry-mac.app/Contents/PkgInfo
```

(the `-` after `nixbld` is `uchg` if I set that flag, `schg,uchg` if I set both of those, etc)

**Comment by edolstra** [comment]
Did you remove the uchg bit on `/nix/store/.links/1666wf33g20gqagij48gmh2wk7vxq07kp0a0avqf13hyjfyii7wx`? That also needs to be removed, since otherwise the link count on that file cannot be changed, I think.

**Comment by clacke** [comment]
I checked, and that file does not exist yet, it's creating it that fails, I suppose.

**Comment by clacke** [comment]
So it's unlikely that your patch helps, there seems to be something else going on. But I'll try it, maybe I made a mistake somewhere along the way.

**Comment by matthewbauer** [comment]
Hmmm... interesting I just found this unanswered stack exchange post:

https://apple.stackexchange.com/questions/202780/why-can-i-not-make-a-hard-link-to-pkginfo

and another link:
https://groups.google.com/a/continuum.io/forum/#!topic/anaconda/CB0zClGfh3c

**Comment by clacke** [comment]
Oh. That's really disappointing. I guess the fix is to not try to optimize files with that name in a directory of that name, when the OS is MacOS. :-/

Is it possible for someone to actually-move an issue, or do I just close this one and reference it from a nix issue?

**Comment by matthewbauer** [comment]
This can be closed along with NixOS/nix#1443

**Comment by clacke** [comment]
Awesome, thanks!
